pkg/security/security_profile/manager.go

```diff
@@ -0,0 +1,2376 @@
+	m.pendingCacheLock.Lock()
+	defer m.pendingCacheLock.Unlock()
+
+	p.InstancesLock.Lock()
+	defer p.InstancesLock.Unlock()
+	// check if the profile should be deleted
+	if len(p.Instances) != 0 {
+		// this profile is still in use, leave now
+		return
+	}
+
+	// remove the profile from the list of profiles
+	delete(m.profiles, *p.GetWorkloadSelector())
+
+	if p.LoadedInKernel.Load() {
+		// remove profile from kernel space
+		m.unloadProfileMap(p)
+		if err := m.persistProfile(p); err != nil {
+			seclog.Errorf("couldn't persist profile: %v", err)
+		}
+	}
+
+	// cleanup profile before insertion in cache
+	p.Reset()
+
+	// TODO: is it possible to have a profile with no selector here?
+	if p.GetWorkloadSelector().IsReady() {
+		// do not insert in cache
+		return
+	}
+
+	// add profile in cache
+	m.pendingCache.Add(*p.GetWorkloadSelector(), p)
+}
+
+// persistProfile (thread unsafe) persists a profile to the filesystem
+func (m *Manager) persistProfile(p *profile.Profile) error {
+	raw, err := p.EncodeSecurityProfileProtobuf()
+	if err != nil {
+		return fmt.Errorf("couldn't encode profile: %w", err)
+	}
+
+	filename := p.Metadata.Name + ".profile"
+	outputPath := path.Join(m.config.RuntimeSecurity.SecurityProfileDir, filename)
+	tmpOutputPath := outputPath + ".tmp"
+
+	// create output directory and output file, truncate existing file if a profile already exists
+	err = os.MkdirAll(m.config.RuntimeSecurity.SecurityProfileDir, 0400)
+	if err != nil {
+		return fmt.Errorf("couldn't ensure directory [%s] exists: %w", m.config.RuntimeSecurity.SecurityProfileDir, err)
+	}
+
+	file, err := os.OpenFile(tmpOutputPath, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, 0400)
+	if err != nil {
+		return fmt.Errorf("couldn't persist profile to file [%s]: %w", outputPath, err)
+	}
+	defer file.Close()
+
+	if _, err := file.Write(raw.Bytes()); err != nil {
+		return fmt.Errorf("couldn't write profile to file [%s]: %w", tmpOutputPath, err)
+	}
+
+	if err := file.Close(); err != nil {
+		return fmt.Errorf("error trying to close profile file [%s]: %w", file.Name(), err)
+	}
+
+	if err := os.Rename(tmpOutputPath, outputPath); err != nil {
+		return fmt.Errorf("couldn't rename profile file [%s] to [%s]: %w", tmpOutputPath, outputPath, err)
+	}
+
+	seclog.Infof("[profile] file for %s written at: [%s]", p.GetSelectorStr(), outputPath)
+
+	return nil
+}
+
+// onNewProfile handles the arrival of a new profile (or the new version of a profile) from a provider
+func (m *Manager) onNewProfile(newProfile *profile.Profile) {
+	m.profilesLock.Lock()
+	defer m.profilesLock.Unlock()
+
+	// a profile loaded from file can be of two forms:
+	// 1. a profile coming from the activity dump manager, providing an activity tree corresponding to
+	//    the selector image_name + image_tag.
+	// 2. a profile coming from the security profile manager, providing an activity tree corresponding to
+	//    the selector image_name, containing multiple image tag versions. Not yet the case, but it will be.
```

Comment:
I'm not sure the comments are still aligned with the code. here we only have a the case of a dump right?

Reply:
Updated the comment with 6434bf45c6f722361b724c91487a36c18f1b46c4